### DISCUSSION TITLE
Behaviour of +HDHR and +EZHD osu!taiko on lazer.

### DISCUSSION THREAD

**Discussion by mangomizer** [ppy/osu#14798] [Q&A]
Based on a wide survey, it was agreed upon that when playing with +HD on osu!taiko lazer, notes should become fully invisible at the same exact % away from the hit target, regardless of mod combination (eg. +EZHD or +HDHR). 

This is not the case on the latest build of lazer and the osu!taiko playerbase unanimously agrees that this needs to change.

We propose that the notes should fully disappear at around 60% away from hit target to edge of screen (when a note is just outside of view).

![image](https://user-images.githubusercontent.com/30021847/133905540-f0d27181-f6e4-441d-b15a-7884acfa7748.png)

Based on the above assumptions, I would like to know what the community thinks on the behaviour of +HDHR (and +EZHD) implementation

Should:
A) All objects should start fading at the exact point in space, and reach 0% opacity when it hits the green line.
Or
B) Objects can start fading at different points, but reach 0% opacity when it hits the green line. If +HD starts fading at 100%, then +HDHR will start fading off screen (186%), and +EZHD will start fading after appearing on the playfield (80%)

Discuss below, and please give reasonings if appropriate.

**Comment by bdach** [comment]
i'm not sure i can understand the explanations above correctly, but would that mean that hard rock being active would change how hidden behaves? if so, then from what i recall that is something that we _really_ want to avoid in general (inter-mod dependencies).

**Reply by mangomizer** [reply]
Option B is what Lazer currently does. (although the green line moves as well).

Fixing the position of the green line is something that is pretty much non-negotiable and needs to be fixed.
Given the above, I would personally prefer Option A, though historically Option B has been used.

**Reply by ReiFan49** [reply]
I tinkered a bit the `TaikoModHidden.cs` file and manage to "fix" when the map **is not on Base SV 1.4** (after mods applied)

![](https://cdn.discordapp.com/attachments/888801439894356028/888870976878706778/unknown.png)

at least this fixes the fadepoint related issue for lazer.

```diff
-    public class TaikoModHidden : ModHidden
+    public class TaikoModHidden : ModHidden, IApplicableAfterBeatmapConversion
     {
         public override string Description => @"Beats fade out before you hit them!";
         public override double ScoreMultiplier => 1.06;
+        public double SliderBasis = 1.4;
```
```diff
             {
                 case DrawableDrumRollTick _:
                 case DrawableHit _:
-                    double preempt = 10000 / MultiplierAt(hitObject.HitObject.StartTime);
+                    double preempt = (10000 * (1.4 / SliderBasis)) / MultiplierAt(hitObject.HitObject.StartTime);
                     double start = hitObject.HitObject.StartTime - preempt * 0.6;
                     double duration = preempt * 0.3;
```
```diff

         public override void ApplyToBeatmap(IBeatmap beatmap)
         {
+            SliderBasis = beatmap.BeatmapInfo.BaseDifficulty.SliderMultiplier;
             controlPointInfo = beatmap.ControlPointInfo;
         }
     }
```
technically speaking, `ApplyToBeatmap` is invoked when the Mod *implements certain interface*. By locking the "base target SV to 1.4", this also allows the client assumption went through smoothly when trying to make an object fades out when approaching the fadepoint.

**Reply by bdach** [reply]
@ReiFan49 i would ask to refrain touching that code before #14778 is merged because that pull is touching the exact same place. it is already known that the preempt time is calculated quite wrongly.

**Reply by smoogipoo** [reply]
The mentioned PR has since been merged. It changes the calculations here quite a bit so try and find the appropriate values again (hopefully by just adjusting the two constants at the top of the class).

**Comment by mangomizer** [comment]
Consider this: Assuming equal scroll speed, should the fading process change depending on if you are using HD or HDHR?
eg. A 180BPM +HDHR map vs a 334.8BPM +HD map. 

If so, you should vote Option B:
If not, you should vote Option A:

**Comment by osutaiko** [comment]
Option A for sure, since B can be abused easily by adjusting the brightness of the monitor. I'm not sure how this method isn't being used if it was B all along on stable though.

**Comment by nl-tata** [comment]
I think that notes should be visible at 100% opacity for 1 full big note's width. The idea that notes should appear on the screen pre-dimmed is unreasonable to me.

I am also thinking that the EZ mod should also possibly make HD easier, since it is called the easy mod for a reason. Or is that what this whole discussion is about... that EZ and HR shouldn't make HD easier or harder. I personally think that HR in its current state makes HD HR completely unreasonable to play.

I hope that next we can fix the unforgivable flashlight mod.
